dpc/src/virtual_machine/virtual_machine.rs

```diff
@@ -22,20 +22,33 @@ use itertools::Itertools;
 use rand::{CryptoRng, Rng};
 use std::sync::Arc;
 
-#[derive(Clone, Debug)]
+#[derive(Clone, Derivative)]
+#[derivative(Debug(bound = "N: Network"))]
```

Comment:
this probably applies to a lot more objects, but do we even need `Debug` (at least the default impl, which might already be unreadable even without the circuits) here?